File: python/pymess/equations/std.py
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, see <http://www.gnu.org/licenses/>.
#
# Copyright (C) Peter Benner, Martin Koehler, Jens Saak and others
#               2009-2018
#

r"""Interface to mess_equation_lyap, mess_equation_glyap, mess_equation_riccati, mess_equation_griccati."""
from pymess.equation import Equation


class EquationGLyap(Equation):
    r"""Interface to mess_equation_lyap and mess_equation_glyap.

    Represents a standard or generalized Lyapunov equation

    .. math::

        A^T X       + X A           &= -B^T B,          \\
        A X         + X A^T         &= -C C^T,          \\
        A^T X E     + E^T X A       &= -B^T B,          \\
        A X E^T     + E X A^T       &= -C C^T.


    arising from a first order system

    .. math::

            E\dot{x} &= Ax + B u,\\
                y    &= C x.

    Parameters
    ----------
    opt: options instance
        :class:`~pymess.options.Options` instance

    a: (n,n) sparse real matrix
        Matrix :math:`A` of Lyapunov equation.

    e: (n,n) sparse real matrix or None
        Matrix :math:`E` of Lyapunov equation.
        If e is None, then :math:`E` of Lyapunov equation will be the identity matrix.

    rhs: (n,p) array_like real matrix or (p,n) array_like real matrix
        The right hand side of Lyapunov equation either :math:`B` or :math:`C`.
        We assume that :math:`p \ll n`.

    Raises
    ------


    See Also
    --------
    :class:`.EquationGRiccati`
    :func:`pymess.lradi.lradi`

    Notes
    -----
    Supported types are ndarray of numpy and csc, coo or csr matrix of scipy.

    References
    ----------


    Examples
    --------
    Further examples :download:`pymess_lradi.py <../../example/pymess_lradi.py>`


    """

    def __init__(self, opt, a, e, rhs):
        r"""Create a :class:`.EquationGLyap` instance"""
        shapea = a.shape
        dim = shapea[0]
        super(EquationGLyap, self).__init__("EquationGLyap", opt, dim)
        self.a = a
        self.rhs = rhs
        self.e = e

    def __delattr__(self, key):
        raise TypeError(
            "Trying to delete attributes of %r is not allowed" % self)


class EquationGRiccati(Equation):
    r"""Interface to mess_equation_riccati and mess_equation_griccati.

    Represents a standard or generalized Riccati equation

    .. math::

        A^T X   + X A       - XBB^T X       &= -C^T C,      \\
        A X     + X A^T     - XC^TC X       &= -BB^T,       \\
        A^T XE  + E^T X A   - E^TXBB^T XE   &= -C^T C,      \\
        A XE^T  + E X A^T   - EXC^TC XE^T   &= -BB^T.

    arising from a first order system

    .. math::

            E\dot{x}&= Ax + B u,   \\
                y   &= C x.

    Parameters
    ----------
    opt: options instance
        :class:`~pymess.options.Options` instance

    a: (n,n) sparse real matrix
        Matrix :math:`A` of Riccati equation.

    e: (n,n) sparse real matrix or None
        Matrix :math:`E` of Riccati equation.
        If e is None, then :math:`E` of Riccati equation will be the identity matrix.


    b: (n,p) array_like real matrix
        Matrix :math:`B` of Riccati equation.
        We assume that  :math:`p \ll n`.

    c: (q,n) array_like real matrix
        Matrix :math:`C` of Riccati equation.
        We assume that  :math:`q \ll n`.


    Raises
    ------


    See Also
    --------
    :class:`pymess.equations.std.EquationGLyap`
    :func:`pymess.lrnm.lrnm`

    Notes
    -----
    Supported types are ndarray of numpy and csc, coo or csr matrix of scipy.

    References
    ----------

    Examples
    --------
    Further examples :download:`pymess_lrnm.py <../../example/pymess_lrnm.py>`

    """

    def __init__(self, opt, a, e, b, c):
        r"""Create a :class:`.EquationGRiccati` instance"""
        shapea = a.shape
        dim = shapea[0]
        super(EquationGRiccati, self).__init__("EquationGRiccati", opt, dim)
        self.a = a
        self.b = b
        self.c = c
        self.e = e
        self.rhs = None

    def __delattr__(self, key):
        raise TypeError(
            "Trying to delete attributes of %r is not allowed" % self)

File: matlab/c_interface/conv/eqn/std.c
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, see <http://www.gnu.org/licenses/>.
//
// Copyright (C) Peter Benner, Martin Koehler, Jens Saak and others
//               2009-2018
//


/**
 * @file matlab/c_interface/conv/eqn/std.c
 * @brief Equation conversion for dae1 function handles.
 * @author @mbehr
 *
 */


#include "interface_matlab.h"

/*-----------------------------------------------------------------------------
 *  Convert a MEX-M.E.S.S. mess_equation_glyap to C-M.E.S.S. mess_equation_glyap
 *-----------------------------------------------------------------------------*/
mess_equation eqn_conv_lyap(const mxArray* m_eqn, const mxArray* m_opt,  mess_freelist mem) {
    int ret = 0;

    /*-----------------------------------------------------------------------------
     *  check input pointer m_eqn and class
     *-----------------------------------------------------------------------------*/
    if(!m_eqn){
        csc_error_message("m_eqn points to NULL.\n");
        return NULL;
    }

    if (!mxIsClass(m_eqn, "mess_equation_glyap")){
        csc_error_message("Expected mess_equation_glyap class.\nAn error occured return NULL Pointer.");
        return NULL;
    }

    /*-----------------------------------------------------------------------------
     *  check input pointer m_opt and class
     *------------------------------------------------------------------------------*/
    if(!m_opt){
        csc_error_message("m_opt points to NULL.\nAn error occured return NULL Pointer.");
        return NULL;
    }

    if (!mxIsClass(m_opt, "mess_options")){
        csc_error_message("Expected mess_options class.\nAn error occured return NULL Pointer.");
        return NULL;
    }


    /*-----------------------------------------------------------------------------
     * check mem pointer
     *-----------------------------------------------------------------------------*/
    if(!mem){
        csc_error_message("mem points to NULL.\n");
        return NULL;
    }

    /*-----------------------------------------------------------------------------
     *  get matrices
     *-----------------------------------------------------------------------------*/
    mess_matrix A       = mess_matrix_from_mexmess(mxGetProperty(m_eqn,0,"A"));
    mess_matrix E       = mess_matrix_from_mexmess(mxGetProperty(m_eqn,0,"E"));
    mess_matrix RHS     = mess_matrix_from_mexmess(mxGetProperty(m_eqn,0,"RHS"));

    /*-----------------------------------------------------------------------------
     *  get options
     *-----------------------------------------------------------------------------*/
    mess_options opt = mess_options_from_mexmess(m_opt);

    /*-----------------------------------------------------------------------------
     *  build equation
     *-----------------------------------------------------------------------------*/
    mess_equation eqn;
    mess_equation_init(&eqn);
    //ret = mess_equation_glyap(eqn, opt, A, E, RHS);
    ret = mess_equation_lyap(eqn, opt, A, E, RHS);

    /*-----------------------------------------------------------------------------
     *  add everything we need to freelist
     *-----------------------------------------------------------------------------*/
    mess_freelist_add_mess_matrix(mem, A);
    mess_freelist_add_mess_matrix(mem, RHS);
    if(E)
        mess_freelist_add_mess_matrix(mem, E);

    mess_freelist_add_mess_equation(mem,eqn);

    /*-----------------------------------------------------------------------------
     *  clear options structure
     *-----------------------------------------------------------------------------*/
    mess_options_clear(&opt);


    return eqn;
}


mess_equation eqn_conv_ric(const mxArray* m_eqn, const mxArray* m_opt,  mess_freelist mem) {
    int ret = 0;

    /*-----------------------------------------------------------------------------
     *  check input pointer m_eqn and class
     *-----------------------------------------------------------------------------*/
    if(!m_eqn){
        csc_error_message("m_eqn points to NULL.\n");
        return NULL;
    }

    if (!mxIsClass(m_eqn, "mess_equation_griccati")){
        csc_error_message("Expected mess_equation_griccati class.\nAn error occured return NULL Pointer.");
        return NULL;
    }

    /*-----------------------------------------------------------------------------
     *  check input pointer m_opt and class
     *------------------------------------------------------------------------------*/
    if(!m_opt){
        csc_error_message("m_opt points to NULL.\nAn error occured return NULL Pointer.");
        return NULL;
    }

    if (!mxIsClass(m_opt, "mess_options")){
        csc_error_message("Expected mess_options class.\nAn error occured return NULL Pointer.");
        return NULL;
    }


    /*-----------------------------------------------------------------------------
     * check mem pointer
     *-----------------------------------------------------------------------------*/
    if(!mem){
        csc_error_message("mem points to NULL.\n");
        return NULL;
    }

    /*-----------------------------------------------------------------------------
     *  get matrices
     *-----------------------------------------------------------------------------*/
    mess_matrix A       = mess_matrix_from_mexmess(mxGetProperty(m_eqn,0,"A"));
    mess_matrix E       = mess_matrix_from_mexmess(mxGetProperty(m_eqn,0,"E"));
    mess_matrix B       = mess_matrix_from_mexmess(mxGetProperty(m_eqn,0,"B"));
    mess_matrix C       = mess_matrix_from_mexmess(mxGetProperty(m_eqn,0,"C"));

    /*-----------------------------------------------------------------------------
     *  get options
     *-----------------------------------------------------------------------------*/
    mess_options opt = mess_options_from_mexmess(m_opt);

    /*-----------------------------------------------------------------------------
     *  build equation
     *-----------------------------------------------------------------------------*/
    mess_equation eqn;
    mess_equation_init(&eqn);
    //ret = mess_equation_griccati(eqn, opt, A, E, B, C);
    ret = mess_equation_riccati(eqn, opt, A, E, B, C);

    /*-----------------------------------------------------------------------------
     *  add everything we need to freelist
     *-----------------------------------------------------------------------------*/
    mess_freelist_add_mess_matrix(mem, A);
    mess_freelist_add_mess_matrix(mem, B);
    mess_freelist_add_mess_matrix(mem, C);
    if(E)
        mess_freelist_add_mess_matrix(mem, E);

    mess_freelist_add_mess_equation(mem,eqn);

    /*-----------------------------------------------------------------------------
     *  clear options structure
     *-----------------------------------------------------------------------------*/
    mess_options_clear(&opt);


    return eqn;
}

